Add coordinate to dimension support in from NetCDF with `collect_attrs=False`

jp-dark opened the pull request.

jp-dark committed:
Implement coords_by_dim for attr_per_array NetCDF grouping

jp-dark committed:
Fix parametrize for NetCDF collect_attr tests

jp-dark committed:
Add second noqa comment for C901 to support both Python 3.7 and 3.9

gsakkis commented on tiledb/cf/engines/netcdf4_engine.py:
```diff
@@ -668,10 +676,9 @@ def _from_group_to_attr_per_array(
                         else tuple(chunks)
                     )
                 array_name = ncvar.name
+                is_sparse = any(dim_name in coord_names for dim_name in coord_names)
```
As currently written, this is equivalent to `bool(coord_names)`

gsakkis commented on tiledb/cf/engines/netcdf4_engine.py:
```diff
@@ -638,27 +638,35 @@ def _from_group_to_attr_per_array(
                 from. Use ``'/'`` to specify the root group.
         """
         converter = cls(default_input_file, default_group_path)
+        coord_names = []
```
This can be a `set()` since it's used only for membership tests

gsakkis commented on tiledb/cf/engines/netcdf4_engine.py:
```diff
@@ -638,27 +638,35 @@ def _from_group_to_attr_per_array(
+                continue
             if not ncvar.dimensions:
                 array_name = (
                     "scalars" if "scalars" not in netcdf_group.variables else "_scalars"
```
This looks ambiguous; if a "scalars" array is added, it may or may not be in the original netcdf. Why not use the underscored name always? Or perhaps this doesn't matter?

jp-dark committed:
Fix which arrays are sparse when ``coords_to_dims=True``

Fix a typo in checking the array has any coordinate dimensions.
Previously, if any coordinates existed in a NetCDF group being
converted, then all arrays would be set to sparse. It should only be
arrays that contain that dimension.

jp-dark committed:
refactor: Change lists for checking coordinate names to sets

jp-dark committed:
Update scalar array name to be something that can be set

jp-dark committed:
refactor: Add function to help get chunks from a NetCDF variable

gsakkis commented on tiledb/cf/engines/netcdf4_engine.py:
```diff
@@ -632,48 +633,56 @@ def _from_group_to_attr_per_array(
+                array_tiles = tiles_by_var.get(
+                    ncvar.name,
+                    tiles_by_dims.get(ncvar.dimensions, get_variable_chunks(ncvar)),
+                )
```
Can `tiles_by_var`/`tiles_by_dims` return a falsy value (`None, (), 0`)? If not I'd rewrite with `or`:

```py
array_tiles = (
    tiles_by_var.get(ncvar.name) or 
    tiles_by_dims.get(ncvar.dimensions) or 
    get_variable_chunks(ncvar)
)

jp-dark replied:
They can return a `None` value.

gsakkis commented on tiledb/cf/engines/netcdf4_engine.py:
```diff
@@ -601,15 +601,16 @@ def from_group(
-    @classmethod
-    def _from_group_to_attr_per_array(
+    @classmethod  # noqa: C901
+    def _from_group_to_attr_per_array(  # noqa: C901
```
Is this still needed after extracting `get_variable_chunks`?

jp-dark replied:
Yeah, it does.

gsakkis commented on tiledb/cf/engines/netcdf4_engine.py:
```diff
@@ -676,7 +676,9 @@ def _from_group_to_attr_per_array(  # noqa: C901
-                is_sparse = any(dim_name in coord_names for dim_name in coord_names)
+                is_sparse = any(
+                    dim_name in coord_names for dim_name in ncvar.dimensions
+                )
```
A test case for this fix would be nice

jp-dark replied:
Added as clubhouse story 8881.
